File: 2187-minimum-time-to-complete-trips/2187-minimum-time-to-complete-trips.java
class Solution {
    public long minimumTime(int[] time, int totalTrips) {
        Arrays.sort(time);
        long low = 0, high = Long.MAX_VALUE; 
        long minTime = Long.MAX_VALUE;
        
        while (low <= high) {
            long mid = low + (high - low) / 2;
            
            int trips = 0;
            for (int t : time) {
                trips += mid / t;
                
                if (trips < 0) {
                    high = mid-1;
                    break;
                }
            }
            
            if (trips < 0)  continue;
            
            if (trips >= totalTrips) {
                minTime = Math.min(mid, minTime);
                high = mid-1;
            } else {
                low = mid+1;
            }
        }
        
        return minTime;
    }
}

File: 2187-minimum-time-to-complete-trips/2187-minimum-time-to-complete-trips.py
class Solution:
    def minimumTime(self, time: List[int], totalTrips: int) -> int:
      lo = 1   # 1
      hi = min(time) * totalTrips # 45

      def tripsCount(trip_time):  # 34
        res = 0
        for i in range(len(time)):
          res += trip_time//time[i]  # 24 // [5,10,10]  -> [4,2,2] -> 8
        return res

      while lo < hi: # [24,34]
        mid = lo + (hi-lo)//2  # 24
        trip_count = tripsCount(mid) # 8

        print(lo, mid, hi, trip_count)


        # if trip_count == totalTrips:
        #   return mid
        if trip_count >= totalTrips:
          hi = mid
        else:
          lo = mid + 1

      return  lo 